File: src/main/java/szewek/mcflux/fluxcompat/addons/TeslaFluxCompat.java
package szewek.mcflux.fluxcompat.addons;

import net.darkhax.tesla.api.ITeslaConsumer;
import net.darkhax.tesla.api.ITeslaHolder;
import net.darkhax.tesla.api.ITeslaProducer;
import net.darkhax.tesla.lib.TeslaUtils;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.ICapabilityProvider;
import szewek.mcflux.U;
import szewek.mcflux.fluxcompat.EnergyType;
import szewek.mcflux.fluxcompat.FluxCompat;
import szewek.mcflux.fluxcompat.ForgeEnergyCapable;
import szewek.mcflux.fluxcompat.LazyEnergyCapProvider;
import szewek.mcflux.util.ErrMsg;
import szewek.mcflux.util.InjectCond;
import szewek.mcflux.util.MCFluxReport;

@FluxCompat.Addon(requires = InjectCond.MOD, args = {"tesla", "TESLA"})
public class TeslaFluxCompat implements FluxCompat.Lookup {
	@Override
	public void lookFor(LazyEnergyCapProvider lecp, FluxCompat.Registry r) {
		final ICapabilityProvider icp = lecp.getObject();
		if (icp == null) return;
		EnumFacing f = null;
		try {
			for (int i = 0; i < U.FANCY_FACING.length; i++) {
				f = U.FANCY_FACING[i];
				if (TeslaUtils.hasTeslaSupport(icp, f)) {
					r.register(EnergyType.TESLA, TeslaFluxCompat::factorize);
				}
			}
		} catch (Exception e) {
			MCFluxReport.addErrMsg(new ErrMsg.BadImplementation("TESLA", icp.getClass(), e, f));
		}
	}

	private static void factorize(LazyEnergyCapProvider lecp) {
		final ICapabilityProvider icp = lecp.getObject();
		Energy[] es = new Energy[7];
		EnumFacing f;
		for (int i = 0; i < U.FANCY_FACING.length; i++) {
			f = U.FANCY_FACING[i];
			es[i] = new Energy(icp, f);
		}
		lecp.update(es, new int[0], null, true);
	}

	private static final class Energy extends ForgeEnergyCapable implements FluxCompat.Convert {
		private final ITeslaHolder holder;
		private final ITeslaConsumer consumer;
		private final ITeslaProducer producer;

		Energy(ICapabilityProvider provider, EnumFacing f) {
			holder = TeslaUtils.getTeslaHolder(provider, f);
			consumer = TeslaUtils.getTeslaConsumer(provider, f);
			producer = TeslaUtils.getTeslaProducer(provider, f);
		}

		@Override public boolean canInputEnergy() {
			return consumer != null;
		}

		@Override public boolean canOutputEnergy() {
			return producer != null;
		}

		@Override public long inputEnergy(long amount, boolean sim) {
			return consumer != null ? consumer.givePower(amount, sim) : 0;
		}

		@Override public long outputEnergy(long amount, boolean sim) {
			return producer != null ? producer.takePower(amount, sim) : 0;
		}

		@Override public long getEnergy() {
			return holder != null ? holder.getStoredPower() : 0;
		}

		@Override public long getEnergyCapacity() {
			return holder != null ? holder.getCapacity() : 0;
		}

		@Override public boolean hasNoEnergy() {
			return holder != null && holder.getStoredPower() == 0;
		}

		@Override public boolean hasFullEnergy() {
			return holder != null && holder.getStoredPower() == holder.getCapacity();
		}

		@Override
		public EnergyType getEnergyType() {
			return EnergyType.TESLA;
		}
	}
}

File: src/main/java/szewek/mcflux/fluxcompat/addons/TeslaFluxCompat.kt
package szewek.mcflux.fluxcompat.addons

import net.darkhax.tesla.api.ITeslaConsumer
import net.darkhax.tesla.api.ITeslaHolder
import net.darkhax.tesla.api.ITeslaProducer
import net.darkhax.tesla.lib.TeslaUtils
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.ICapabilityProvider
import szewek.fl.energy.IEnergy
import szewek.mcflux.U
import szewek.mcflux.fluxcompat.EnergyType
import szewek.mcflux.fluxcompat.FluxCompat
import szewek.mcflux.fluxcompat.ForgeEnergyCapable
import szewek.mcflux.fluxcompat.LazyEnergyCapProvider
import szewek.mcflux.network.CloudUtils
import szewek.mcflux.util.ErrMsg
import szewek.mcflux.util.InjectCond
import szewek.mcflux.util.MCFluxReport

@FluxCompat.Addon(requires = InjectCond.MOD, args = ["tesla", "TESLA"])
class TeslaFluxCompat : FluxCompat.Lookup {
	override fun lookFor(lecp: LazyEnergyCapProvider, r: FluxCompat.Registry) {
		val icp = lecp.obj ?: return
		var f: EnumFacing? = null
		try {
			for (i in 0 until U.FANCY_FACING.size) {
				f = U.FANCY_FACING[i]
				if (TeslaUtils.hasTeslaSupport(icp, f)) {
					r.register(EnergyType.TESLA, ::factorize as FluxCompat.Factory)
				}
			}
		} catch (e: Exception) {
			MCFluxReport.addErrMsg(ErrMsg.BadImplementation("TESLA", icp.javaClass, e, f))
		}

	}

	private fun factorize(lecp: LazyEnergyCapProvider) {
		val icp = lecp.obj
		val es = arrayOfNulls<IEnergy>(7)
		for (i in 0 until U.FANCY_FACING.size) {
			es[i] = Energy(icp, U.FANCY_FACING[i])
		}
		lecp.update(es, IntArray(0), null, true)
		CloudUtils.reportEnergy(icp!!.javaClass, (es[0] as Energy).holder!!.javaClass, "tesla")
	}

	private class Energy internal constructor(provider: ICapabilityProvider?, f: EnumFacing?) : ForgeEnergyCapable(), FluxCompat.Convert {
		internal val holder: ITeslaHolder? = TeslaUtils.getTeslaHolder(provider, f)
		private val consumer: ITeslaConsumer? = TeslaUtils.getTeslaConsumer(provider, f)
		private val producer: ITeslaProducer? = TeslaUtils.getTeslaProducer(provider, f)

		override val energyType
			get() = EnergyType.TESLA

		override fun canInputEnergy() = consumer != null

		override fun canOutputEnergy() = producer != null

		override fun inputEnergy(amount: Long, sim: Boolean) = consumer?.givePower(amount, sim) ?: 0

		override fun outputEnergy(amount: Long, sim: Boolean) = producer?.takePower(amount, sim) ?: 0

		override fun getEnergy() = holder?.storedPower ?: 0

		override fun getEnergyCapacity() = holder?.capacity ?: 0

		override fun hasNoEnergy() = holder != null && holder.storedPower == 0L

		override fun hasFullEnergy() = holder != null && holder.storedPower == holder.capacity
	}
}